### PR TITLE
Add new platform: tx4810-dn

### PR DESCRIPTION
**Pull request by chenglin-tsai**
The tx4810-dn is also built for 3rd party developers. Except the model name, it is all the same to tx4810 platform.

Test log:
[dentOS_tx4810_dn_onlp_test_0114_2021.log](https://github.com/dentproject/dentOS/files/5812598/dentOS_tx4810_dn_onlp_test_0114_2021.log)

Signed-off-by: chenglin-tsai <chenglin.tsai@deltaww.com>

### CODE REVIEW

File: packages/platforms/delta/arm64/tx4810-dn/modules/builds/src/arm64-delta-tx4810-dn-led.c [hunk 1]
```diff
@@ -0,0 +1,391 @@
+/*
+ * A LED driver for the delta_tx4810_dn_led
+ *
+ * Copyright (C) 2021 Delta Networks, Inc.
+ *
+ * Chenglin Tsai <chenglin.tsai@deltaww.com>
+ *
+ * Base on:
```
Comment (paulmenzel) [inline packages/platforms/delta/arm64/tx4810-dn/modules/builds/src/arm64-delta-tx4810-dn-led.c:8]: Base*d* on?

File: packages/base/any/kernels/5.8-lts/patches/0014-delta-tx4810-dn-dts.patch [hunk 2]
```diff
@@ -0,0 +1,171 @@
+--- a/arch/arm64/boot/dts/marvell/delta-tx4810-dn.dts
```
Comment (paulmenzel) [inline packages/base/any/kernels/5.8-lts/patches/0014-delta-tx4810-dn-dts.patch:1]: Please add a git commit message, and also sign it of. Is this in Linux upstream already?

**Review by paulmenzel — COMMENTED** [review]
Thank you for the changes. It’d be great, if it was more clear, where patches come from. Also, is there a way to first import the drivers (from ONL, what commit?) and then do your changes in a separate commit, so it’s clear, what you did, and people do not have to diff it manually.

**Comment by chenglin-tsai** [comment]
> Thank you for the changes. It’d be great, if it was more clear, where patches come from. Also, is there a way to first import the drivers (from ONL, what commit?) and then do your changes in a separate commit, so it’s clear, what you did, and people do not have to diff it manually.

File: packages/platforms/delta/arm64/tx4810-dn/modules/builds/src/arm64-delta-tx4810-dn-led.c (hunk 1 above)
Comment (chenglin-tsai) [reply packages/platforms/delta/arm64/tx4810-dn/modules/builds/src/arm64-delta-tx4810-dn-led.c]: You are right. Fixed

File: packages/base/any/kernels/5.8-lts/patches/0014-delta-tx4810-dn-dts.patch (hunk 2 above)
Comment (chenglin-tsai) [reply packages/base/any/kernels/5.8-lts/patches/0014-delta-tx4810-dn-dts.patch]: This dts is not in the Linux upstream, so I changed the source to /dev/null, and hope it is more clear now.

**Comment by chenglin-tsai** [comment]
> Also, is there a way to first import the drivers (from ONL, what commit?) and then do your changes in a separate commit, so it’s clear, what you did, and people do not have to diff it manually.

Good question. Because the drivers (led driver, cpld driver and so on) in onlp are platform specific and are related to platform hardware design, it will be more clear if the drivers of each platform are independent to each other.

For example, in tx4810-dn led driver, we use the below gpio pins to control the led of PSU/FAN/SYS. But in other platforms, it might be using different gpio pins or even using cpld registers or other ways to control the led. 

```
static struct tx4810_dn_gpio tx4810_dn_led_gpio[] = {
    {"LED PSU Green", 35},
    {"LED PSU Amber", 34},
    {"LED FAN Green", 46},
    {"LED FAN Amber", 43},
    {"LED SYS Green", 63},
    {"LED SYS Amber", 66},
};
```

**Comment by chenglin-tsai** [comment]
Update the test log based on https://github.com/dentproject/dentOS/pull/29/commits/289dcc7663e43a5e996be07a7bd7cb85a69eafb5:

[dentOS_tx4810_dn_onlp_test_0115_2021.log](https://github.com/dentproject/dentOS/files/5818787/dentOS_tx4810_dn_onlp_test_0115_2021.log)

**Comment by chenglin-tsai** [comment]
Hi @paulmenzel 
Do you have any updates to this pull request? And feel free to let me know if anything was missed.

File: packages/platforms/delta/arm64/tx4810-dn/modules/builds/src/arm64-delta-tx4810-dn-led.c [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.     See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+ */
```
Comment (paulmenzel) [inline packages/platforms/delta/arm64/tx4810-dn/modules/builds/src/arm64-delta-tx4810-dn-led.c:26]: Linux uses SPDX style headers now.

File: packages/platforms/delta/arm64/tx4810-dn/modules/builds/src/arm64-delta-tx4810-dn-led.c [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+#include <linux/leds.h>
+#include <linux/slab.h>
+#include <linux/gpio.h>
+#include <linux/bitops.h>
+
+
+#define DRVNAME "tx4810_dn_led"
+
+/* LED related data
+ */
```
Comment (paulmenzel) [inline packages/platforms/delta/arm64/tx4810-dn/modules/builds/src/arm64-delta-tx4810-dn-led.c:42]: Put it on one line?

File: packages/platforms/delta/arm64/tx4810-dn/modules/builds/src/arm64-delta-tx4810-dn-led.c [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+enum led_type {
+    LED_TYPE_PSU,
+    LED_TYPE_FAN,
+    LED_TYPE_SYS,
+    NUM_OF_LED
+};
```
Comment (paulmenzel) [inline packages/platforms/delta/arm64/tx4810-dn/modules/builds/src/arm64-delta-tx4810-dn-led.c:48]: Although a Linux kernel module, the whole file does not follow the [Linux kernel coding style](https://www.kernel.org/doc/html/latest/process/coding-style.html).

File: packages/platforms/delta/arm64/tx4810-dn/modules/builds/src/arm64-delta-tx4810-dn-led.c [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+
+
+static enum led_light_mode gpio_val_to_led_mode(int led_green, int led_amber)
+{
+    if (led_green && !led_amber)
+        return LED_MODE_GREEN;
+    else if (!led_green && led_amber)
+        return LED_MODE_AMBER;
+    else
+        return LED_MODE_OFF;
```
Comment (paulmenzel) [inline packages/platforms/delta/arm64/tx4810-dn/modules/builds/src/arm64-delta-tx4810-dn-led.c:93]: I’d print a warning, if the function is called with both values set (≠ 0).

File: packages/platforms/delta/arm64/tx4810-dn/modules/builds/src/arm64-delta-tx4810-dn-led.c [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+    mutex_lock(&ledctl->update_lock);
+
+    if (time_after(jiffies, ledctl->last_updated + HZ + HZ / 2)
+        || !ledctl->valid) {
+        int i;
+
+        dev_dbg(&ledctl->pdev->dev, "Starting tx4810_dn_led update\n");
+
+        /* Update LED data
+         */
```
Comment (paulmenzel) [inline packages/platforms/delta/arm64/tx4810-dn/modules/builds/src/arm64-delta-tx4810-dn-led.c:167]: Should be one line?

**Review by paulmenzel — COMMENTED** [review]
Thank you for the update.

Please squash the fix-up commit into the first commit.

I am still not happy about the Linux drivers, but, note, I have *no* authority in the project, so take my comments with a grain of salt.

The best way would be to send the drivers to the Linux kernel, also to get them properly reviewed. To make that easily possible in the future, the files should follow the Linux kernel coding style.

The situation in ONL is already a mess, so the question is, if DENT wants to fix that or not.

I’d copy drivers from ONL or other projects as below:

1.  Commit to copy the driver with the commit stating, where it’s from
2.  Commit with minimal changes to get the driver to build (not to work) with appropriate warning in the commit message
3.  Commit the formatting of the file to adhere to the Linux kernel coding style
4.  Commit with the actual changes

PS: When and where can I buy this switch? ;-)

**Comment by chenglin-tsai** [comment]
Hi @paulmenzel,

Thank you for theses suggestions. 

I will fix the CPLD and LED drivers to Linux kernel coding style in the next commit. 

However, I am not sure whether these two drivers could be merged into Linux Kernel, since they are platform drivers and have much relation to platform's hardware layout and definitions, which means they can only work on this platform and are not reusable by others. Like the LED driver in this platform, it is not a real hardware device, it is just composed by few gpio pins of CPU and make it can control the color of front panel LEDs. 

Do you have any example of such kind of platform driver that had merged into Linux Kernel? I would like to know how thing are going.

For above reasons, I could understand why the software architecture of ONL/DENT are designed to let each platform to implement their own drivers.

PS: I will write an email to you to see how to help you get this switch.

**Comment by chenglin-tsai** [comment]
Hi @paulmenzel,
For some reasons, I recreated a new pull request for tx4810-dn platform, and rewrote the led and cpld driver to Linux coding style based on your suggestions.

Please let me know if you still have any concerns.

File: packages/platforms/delta/arm64/tx4810-dn/modules/builds/src/arm64-delta-tx4810-dn-led.c (hunk 7 above)
Comment (chenglin-tsai) [reply packages/platforms/delta/arm64/tx4810-dn/modules/builds/src/arm64-delta-tx4810-dn-led.c:167]: The above suggestions are all taken into led and cpld driver. Hope it is not too late. Thanks a lot.

**Comment by paulmenzel** [comment]
Thank you for improving the changes. I’d preferred to continue using this merge/pull request, but https://github.com/dentproject/dentOS/pull/37 is fine too I guess.

I just wanted to comment regarding upstreaming. Even if the driver is device specific, they should still be upstreamed. This decreases maintenance, and the Linux kernel already contains device specific files, for example the (ARM) devicetree files. Lastly, I guess for a lot of drivers, a generic driver could be implemented, only taking some configuration struct for this specific device. Maybe that could also be added to the devicetree.

**Pull request closed without merging**